merging backend logic

amjedidiah opened the pull request.

amjedidiah committed:
feat(security): implemented pre-signed URL upload approach

BREAKING CHANGE: Replaced direct cloud storage configuration with pre-signed URLs

- Added pre-signed URL endpoint configuration
- Removed cloud storage credentials from frontend
- Deleted cloud storage config files
- Updated upload logic to use pre-signed URLs
- Simplified configuration interface

amjedidiah committed:
fix(upload): resolved CORS issues in pre-signed URL upload

- Added CORS mode to uploadObject fetch request
- Prettier fixes

BREAKING CHANGE: Requires S3 bucket CORS configuration

amjedidiah committed:
fix(app): build and prettier fixes

amjedidiah committed:
chore(app): undoing unnecessary commit changes

amjedidiah committed:
feat(storage): implemented and refined cloud storage SDK integrations

- Added client-side SDK approach for cloud storage providers
- Implemented AWS integration with presigned URLs
- Added credential refresh and SAS token management
- Implemented progress tracking and retry mechanisms
- Fixed type errors and improve error handling
- Removed unused code and align with latest SDK versions

BREAKING CHANGE: Storage configuration requires provider-specific SDK setup

amjedidiah committed:
feat(storage): implemented Azure SDK integrations

- Added client-side SDK approach for Azure provider
- Implemented Azure integration with presigned URLs
- Implemented progress tracking
- Removed unused code and align with latest SDK versions

amjedidiah committed:
chore(app): removed todo from commit

amjedidiah committed:
refactor(app): streamline storageConfig prop values

- Storage config only requires `provider` and `tokenEndpoint` values

amjedidiah committed:
feat(storage): implemented DO SDK integrations

- Added client-side SDK approach for DO provider
- Implemented DO integration with presigned URLs

amjedidiah committed:
feat(app): s3 compliant support

- Unified s3 client-side upload approach for S3 complaint storage providers

amjedidiah committed:
feat(app): merged azure and s3 SDK into one

- preview url for uploaded files

amjedidiah committed:
fix(app): fixed console warning about ETag header

- made ETag header conditional
- removed `previewUrl` from object returned to FE

amjedidiah committed:
feat(app): backend logic

- changed folder structure to accommodate BE logic

amjedidiah committed:
chore(app): build fixes and clean up

amjedidiah committed:
chore(app): better path resolutions and prettier fixes

amjedidiah committed:
feat(app): dual builds

amjedidiah committed:
feat(app): successful merged builds

amjedidiah committed:
fix(app): prettier fixes

BSalaeddin commented on src/frontend/lib/uploadObject.ts:
```diff
@@ -0,0 +1,53 @@
+            method: 'PUT',
+            body: fields ? formData : file,
+            headers: {
+                'Content-Type': file.type,
```
When sending FormData, manually setting the 'Content-Type' header interferes with the automatic setting of the boundary parameter, causing the request to fail.

Recommendation: Remove the 'Content-Type' header when fields are present and FormData is used. Let the browser set it automatically. Modify the code as follows:

```
const response = await fetch(presignedUrl, {
    method: 'PUT',
    body: fields ? formData : file,
    headers: fields
        ? undefined // Do not set 'Content-Type' header when sending FormData
        : {
              'Content-Type': file.type,
          },
    mode: 'cors',
    credentials: 'include',
})

```

amjedidiah replied:
Thanks @BSalaeddin , done

BSalaeddin commented on src/frontend/lib/storage/provider.ts:
```diff
@@ -0,0 +1,204 @@
+        return true
+    }
+
+    private handleError(error: unknown): never {
```
Issue: The method relies on checking substrings in err.message to determine the error type. This approach is fragile and can break if error messages change or are localized.

Recommendation: Use more structured error handling by checking specific error properties like name, code, or custom properties returned from the backend. For example, if the backend returns an error object with a code property, you can use that to determine the error type:


private handleError(error: unknown): never {
    if (error instanceof UploadError) throw error;

    const err = error as any;
    if (err.code === 'UNAUTHORIZED') {
        throw new UploadError(
            'Unauthorized access to Provider',
            UploadErrorType.PERMISSION_ERROR
        );
    }

    if (err.code === 'EXPIRED_URL') {
        throw new UploadError(
            'Presigned URL has expired',
            UploadErrorType.EXPIRED_URL,
            true
        );
    }

    throw new UploadError(
        `Upload failed: ${err.message}`,
        UploadErrorType.UNKNOWN_UPLOAD_ERROR,
        true
    );
}

amjedidiah replied:
Thanks @BSalaeddin , done

BSalaeddin commented on src/frontend/lib/storage/provider.ts:
```diff
@@ -0,0 +1,204 @@
+        })
+    }
+
+    validateConfig(): boolean {
```
Validation of Required Configurations:

Issue: The validateConfig method only checks for tokenEndpoint, but provider is also critical for the functionality.

Recommendation: Update the validateConfig method to include provider:

validateConfig(): boolean {
    const required = ['tokenEndpoint', 'provider'] as const;
    const missing = required.filter(key => !this.config[key]);

    if (missing.length > 0) {
        throw new Error(`Missing required configuration: ${missing.join(', ')}`);
    }

    // Optionally, validate that provider is a valid enum value
    if (!Object.values(Provider).includes(this.config.provider)) {
        throw new Error(`Invalid provider: ${this.config.provider}`);
    }

    return true;
}

amjedidiah replied:
Thanks @BSalaeddin  done

BSalaeddin commented on src/frontend/lib/storage/provider.ts:
```diff
@@ -0,0 +1,204 @@
+            })
+
+            xhr.addEventListener('error', () =>
+                reject(new Error('Network error during upload')),
```
Recommendation: Enhance error handling by including xhr.status and xhr.statusText:

```
xhr.addEventListener('error', () =>
    reject(
        new Error(
            `Network error during upload: ${xhr.status} ${xhr.statusText}`,
        ),
    ),
);

```

amjedidiah replied:
Thanks @BSalaeddin , done

amjedidiah committed:
fix(app): review fixes

- undid setting `Content-Type ` header when sending fetch request from `uploadObject`
- improved logic for `handleError`
- improved validation of required configuration
- Enhanced error handling by including `xhr.status` and `xhr.statusText` when reporting error from `uploadWithProgress`
- fixed issue with build failing when temp dist directory exists

amjedidiah committed:
Merge branch 'master' into feat/be-logic
